SR with precomputed gradients

attila-i-szabo opened the pull request:
I have implemented the SR algorithm with precomputed gradients. This is less elegant than the lazy `vjp/jvp`-based implementation in `LazySMatrix`, but has practical advantages:

- It has minimal memory overhead: precomputing the gradients requires the same amount of memory as a single pass of `vjp` since the different samples on which the neural network is run are independent, so we only need to compute one of them to get a row of the Jacobian. (I.e., instead of looping through `vjp` with `vmap` as `jacrev` does, we can loop through `grad`.) Storing the matrix of gradients is guaranteed to take up less memory: backpropagation has to store a lot of internal information about the forward pass, which takes up many times the memory needed for the gradients in a deep network (In my experiments, they were 40 MB vs 1.2 GB). The bottom line is that if there is enough memory for a single `vjp`, there is enough memory for this too.
- It yields a massive speedup: the gradient matrix can again be calculated in the same asymptotic time as a single `vjp` (it will of course be a bit slower because `vmap` is used, but not by large factors); afterwards, we only need matrix-vector multiplications, which are way faster than a full backpropagation of a complex neural network. In my experiments (on 20 CPUs), I could reduce the time of an SR step from 8x that with Adam to about 1.2x.
- It allows for regularising the S matrix in a scale-invariant way by factoring out the magnitude of diagonal elements, as described in Becca & Sorella, p. 143. This has minimal overhead and can be crucial for heterogeneous networks that may have very different gradients in different parts.

Unfortunately, calculating the full Jacobian cannot be done as dtype-agnostically as a VJP, so some information about the network structure needs to be passed. This is done through the `jacobian` parameter, which is implemented for the values `"R2R"` (both the network entries and the wave function are real), `"R2C"` (real entries, complex wave function), and `"holomorphic"` (holomorphic function of complex parameters), `None` (uses a `LazySMatrix`). 
- The parameter could use a better name, but `jacobian` also signifies it is a switch for using this code vs. the original one
- It may be possible to automate this choice, although it's not trivial (e.g., R2R is the right choice even if the wave function has negative entries, which makes the output complex...)
- A few more cases might be implemented, e.g. non-holomorphic C2C, although that is just sugar for R2C...

The boolean parameter `rescale_shift`specifies whether the scale-invariant regularisation should be used.

gcarleo commented:
thanks a lot for this contribution, this is indeed something needed. we'll get back to you with a few comments asap!

PhilipVinc commented on netket/optimizer/sr/sr_onthefly.py:
```diff
@@ -282,3 +300,199 @@ def fun(W, σ):
+    if rescale_shift:
+        O, scale = gradients(apply_fun, params, samples, model_state,
+                             mode, True)
+        np.savetxt('dum', 1/scale[:,jnp.newaxis])
```
leftover?

attila-i-szabo replied:
The print statement is a leftover indeed. I've remembered it just as I left from work... the branching is of course necessary to properly handle scale-invariant regularisation.

PhilipVinc commented on netket/optimizer/sr/sr_onthefly.py:
```diff
@@ -282,3 +300,199 @@ def fun(W, σ):
+        Returns:
+            A dense matrix representation of this S matrix.
+        """
+        return jnp.transpose(jnp.conj(self.O)) @ self.O \
+            + self.sr.diag_shift * jnp.eye(self.O.shape[1])
```
```suggestion
        return self.O.T.conj() @ self.O + self.sr.diag_shift * jnp.eye(self.O.shape[1])
```
is more readable

PhilipVinc commented:
HI @attila-i-szabo ,
Thanks for your contribution! I started giving it a look and it looks very good!
Indeed this is something that was missing and it's a great thing to add it.

However, one of the things on my roadmap before the final v3.0 is to reformulate a bit the interface for constructing the S matrix. So if you do not disagree we will use the occasion to discuss a bit the interface for creating S matrices and using SR, if that's ok for you.

Right now there is some overlap between the S object and algorithm used to compute it, and the algorithm used to solve the linear system Sx=F, leading to a somewhat more messy implementation. You can see this because you had to add a flag to chose the type of S matrix to `SRLazy`.

Once the interface is in place, hopefully rather quickly, we can merge your changes on top of it, if that's ok by you.

In the meantime, to ease the rebasing of changes that will surely happen I'd ask you to: 
- move all your new structures and functions from `sr_onthefly.py` (which is the name we use for the specific implementation with jax.jvp/vjp) to another file. `sr_jacobian.py` seems a good candidate.
- explain me the following: right now the only thing blocking us to use the same `solve` methods of `sr_onthefly` for your new S matrix is that there are a bunch of checks for `S.scale`, which I guess is related to the method you mentioned above. Is there no way to include this into the matrix-vector product so that we can use more generic code for the iterative solver?

attila-i-szabo commented:
Hi, thanks for pointing out your refactoring the SR interface. I'll make sure to have a look at it. 

Re checks on `S.scale`, the scale-invariant regularisation is usually implemented by factoring out the overall scale of rows and columns into scale factors that are applied to the right hand side and the solution of the linear equation. The alternative would be to scale the regularising terms we add to the diagonal of the S matrix with these factors. The way I implemented is preferable in that the standard (identity) preconditioner is an optimal choice for a matrix with only 1s on the diagonal, which you get in this implementation.

PhilipVinc commented:
We can also merge your PR with a little more work on it, and later I'll take care of supporting the new API, I'd just stall you a bit while we discuss it further.

About the scaling... Isn't scaling the regularising term essentially equivalent to what you are doing in the end?

attila-i-szabo commented:
I've just split the new code into its own file so much as possible (it still has to be hooked up with `SRLazy`)

Re scaling: you're right, it effectively multiplies the diagonal entries with 1+epsilon, rather than adding epsilon to it. The advantage of my implementation is that it also takes care of preconditioning: once the scale factors are factored out, all diagonal entries of the S matrix are 1, so the default identity preconditioner works as well as it can. We could fix that by defining a preconditioner that applies the diagonal entries, but then the preconditioner needs to be constructed by the `JacobianSMatrix` and not supplied by the `SRLazy` object.

Another point I wanted to flag: the S matrix for real entries is Re(O^dagger O): is this implemented by the `LazySMatrix`?

inailuig commented on netket/optimizer/sr/sr_jacobian.py:
```diff
@@ -0,0 +1,223 @@
+        
+    return unravel(
+        jnp.transpose(jnp.conj( jnp.transpose(jnp.conj(S.O @ vec)) @ S.O)) +
+        S.sr.diag_shift * vec)
```
Before or after merging we should also add support for MPI.
If you want take up this task then have a look at how this was done in our legacy code [here](https://github.com/netket/netket/blob/4a3abfc0c88092f472a7e87ab89e6a829f5f5776/netket/legacy/optimizer/jax/stochastic_reconfiguration.py#L25) .
Each rank has its own set of samples, and so its own slice of rows of the jacobian, and essentially all that is needed is to take the mean with MPI of the result of the matrix-vector product here.
Also  the scale-inveriant regularization would have to be paralellized with MPI.

gcarleo replied:
I agree this is a very important point

PhilipVinc replied:
good catch

attila-i-szabo replied:
I personally had a hard time with getting MPI to work, it always appeared to me that the different cores just do their own thing without communication. Is there a tutorial on how to actually make MPI to work?

PhilipVinc replied:
https://www.netket.org/docs/getting_started.html#mpi

I should write more docs on it. You can check that MPI is detected by netket by printing (nk.utils.mpi_available).
For the moment you can simply try to `pip install -U 'netket[mpi]'` and then do `mpirun -np X python Examples/Ising1d/ising.py`

gcarleo commented:
> Another point I wanted to flag: the S matrix for real parameters is Re(O^dagger O): is this implemented by the LazySMatrix?

I hope so! :) this should be taken care of already in the existing code, @inailuig ?

PhilipVinc commented:
> Another difference in my solve routine is that the Jacobian is stored as a matrix rather than a pytree, so it makes a lot of sense to ravel the RHS pytree and unravel the result vector, rather than using a JacobianSMatrixx Pytree multiplication that does this at every single iteration.

This is indeed a valid point which we should consider in the redesign of the API, @gcarleo . The issue being that encodings such as this one want the F pytree to be converted to a dense vector eagerly, then solve the linear system, then go back to pytree.
However that is not the default behaviour of what I was proposing, where everything is always converted back to a PyTree after every application of S on F...

PhilipVinc commented:
also @attila-i-szabo there are a few tests you should add in `Tests/Optimizer/sr_onthefly.py` .

attila-i-szabo commented:
> This is indeed a valid point which we should consider in the redesign of the API, @gcarleo . The issue being that encodings such as this one want the F pytree to be converted to a dense vector eagerly, then solve the linear system, then go back to pytree.
> However that is not the default behaviour of what I was proposing, where everything is always converted back to a PyTree after every application of S on F...

I mentioned these issues over at #649 too. I don't see an obvious way around this other than a `solve` routine in the matrix object, which could take care of these differences (and more, such as preconditioners). They could take the solver as an argument to make things more transparent. Also, if we choose to make the different SMatrices subclass of an `SMatrix` class, that could be supplied with a default implementation that just calls the solver with whatever kwargs are passed to it.

inailuig commented:
> > Another point I wanted to flag: the S matrix for real parameters is Re(O^dagger O): is this implemented by the LazySMatrix?
> 
> I hope so! :) this should be taken care of already in the existing code, @inailuig ?

Yes, this is automagically taken care of by casting the final result [here](https://github.com/netket/netket/blob/4a3abfc0c88092f472a7e87ab89e6a829f5f5776/netket/optimizer/sr/sr_onthefly_logic.py#L120). In the $R->C$ case you are referring to the intermediary results is complex, the final result too but with zero imaginary part which we cast away.

attila-i-szabo commented:
> > > Another point I wanted to flag: the S matrix for real parameters is Re(O^dagger O): is this implemented by the LazySMatrix?
> > 
> > 
> > I hope so! :) this should be taken care of already in the existing code, @inailuig ?
> 
> Yes, this is automagically taken care of by casting the final result [here](https://github.com/netket/netket/blob/4a3abfc0c88092f472a7e87ab89e6a829f5f5776/netket/optimizer/sr/sr_onthefly_logic.py#L120). In the R->C case you are referring to the intermediary result is complex, the final result too but with zero imaginary part which we cast away.

Neat, thanks for pointing out!

PhilipVinc commented:
@attila-i-szabo can you rebase this PR on top of #650 and make JacobianSMatrix a subclass of `AbstractSMatrix`?

attila-i-szabo commented:
Hi, I rebased this branch to #650 and added the changes in you gist (thanks for that!). Other than bug fixes, I'll need to tend to the issue of scale factors later today. For now, when `rescale_shift` is on, `@` multiplies the vector with the rescaled S matrix, so i wouldn't work directly with solvers, only the way it is wrapped in `solve`. There are two routes for fixing that:

- Keep the representation of the matrix and the `solve` method as it is (which has the benefit that the identity preconditioner works well) but correct `__matmul__` with the scale factors
- Don't rescale `O` and supply the scale factors in `solve` as a preconditioner and in the diagonal shift in `__matmul__`

Which of these would be preferable?

PhilipVinc commented on netket/optimizer/sr/s_jacobian_mat.py:
```diff
@@ -0,0 +1,205 @@
+    the field `sr`.
+    """
+
+    sr: SRJacobian
+    """Parameters for the solution of the system."""
```
```suggestion
```

That field is in the base class already.

PhilipVinc commented on netket/optimizer/sr/s_jacobian_mat.py:
```diff
@@ -0,0 +1,205 @@
+    for all samples σ_i at given values of the parameters p_j
+    Average <O_j> subtracted for each parameter
+    Divided through by sqrt(#samples) to normalise S matrix
+    If scale is not None, normalised to unit magnitude"""
```
```suggestion
    If scale is not None, normalised to unit magnitude
    """
```

PhilipVinc commented on netket/optimizer/sr/s_jacobian_mat.py:
```diff
@@ -0,0 +1,205 @@
+        else:
+            unravel = lambda x: x
+
+        return unravel(
```
You should only unravel the output if you ravelled the input. 
Otherwise this won't work with scipy solvers.
Check the implementation of onthefly.

attila-i-szabo replied:
In that case, `unravel` is just an identity function (see line 60), but I can change it if confusing

PhilipVinc replied:
ah, yes, sorry.
It's a bit confusing but might be ok

PhilipVinc commented on netket/optimizer/sr/s_jacobian_mat.py:
```diff
@@ -0,0 +1,205 @@
+    vectorising the loop using vmap
+    If rescale_shift is True, columns of O are rescaled to unit magnitude, and
+    scale factor [1/sqrt(S_kk)] returned as a separate vector for
+    scale-invariant regularisation as per Becca & Sorella p. 143."""
```
```suggestion
    scale-invariant regularisation as per Becca & Sorella p. 143.
    """
```

And could you also give more details about what `mode` does? what changes between R2R, R2C and HOLO?

attila-i-szabo replied:
I've expanded the docstring of `SRJacobian.mode`, please check if that's what you had in mind

PhilipVinc commented on netket/optimizer/sr/s_jacobian_mat.py:
```diff
@@ -0,0 +1,205 @@
+    """Calculates the gradient of a neural network for a number of samples
+    efficiently using vmap(grad),
+    subtracts their mean for each parameter, i.e., each column,
+    and divides through with sqrt(#samples) to normalise the S matrix"""
```
```suggestion
    and divides through with sqrt(#samples) to normalise the S matrix
    """
```

PhilipVinc commented on netket/optimizer/sr/s_jacobian_mat.py:
```diff
@@ -0,0 +1,205 @@
+        return grads
+
+
+@partial(jax.jit, static_argnums=(0, 3))
```
no need to jit this if you call it always from jitted function

inailuig commented:
> * It has minimal memory overhead: precomputing the gradients requires the same amount of memory as a single pass of `vjp` since the different samples on which the neural network is run are independent, so we only need to compute one of them to get a row of the Jacobian. (I.e., instead of looping through `vjp` with `vmap` as `jacrev` does, we can loop through `grad`.) Storing the matrix of gradients is guaranteed to take up less memory: backpropagation has to store a lot of internal information about the forward pass, which takes up many times the memory needed for the gradients in a deep network (In my experiments, they were 40 MB vs 1.2 GB). The bottom line is that if there is enough memory for a single `vjp`, there is enough memory for this too.

Thanks for bringing this up :)
The memory needed to store the forward when doing one big vjp (the way it is currently implemented in the master branch) is indeed something we were suspecting to be an issue (not the case for our RBM toy model though). 

My proposed solution is to use the same approach already used for parallelizing it with MPI. The jacobian is split in blocks of rows which are multiplied from the left with a corresponding slice of the row vector. Then you loop over them while summing the results. (Taking it to the extreme you do just a batch of one sample at a time, which is just a vjp of a scalar function with a scalar, which is exactly what jax.grad is doing internally...). This way the memory needed can be reduced to that of a grad of a single sample at the expense of performance.

To anyone interested, feel free to take a look at https://github.com/inailuig/netket/commit/0cd0b56408bafd43eee976e780e704c2280365fb where  I had a go at implementing this a couple of weeks ago (I also batched the jvp to see if it might improve compile times). It should already work for real parameters, but since jax scan/loops currently cannot be transposed (with AD; a requirement by custom_linear_solve in the iterative solvers) this is pending on me getting around to opening an issue/pr at  jax...

Also there would be [jax.checkpoint](https://jax.readthedocs.io/en/latest/jax.html?highlight=jax.checkpoint#jax.checkpoint).

The main motivation for doing SR with jvp/vjp in the first place was that you also already cannot store the gradients (oks) when you have, say 100k parameters and samples. 
However your deep networks seem to be in the regime where precomputing the gradients is preferable, so this contribution is highly welcome!

> * It yields a massive speedup: the gradient matrix can again be calculated in the same asymptotic time as a single `vjp` (it will of course be a bit slower because `vmap` is used, but not by large factors); afterwards, we only need matrix-vector multiplications, which are way faster than a full backpropagation of a complex neural network. In my experiments (on 20 
CPUs), I could reduce the time of an SR step from 8x that with Adam to about 1.2x.

You could try `SR(..., centered=False)` which uses the implementation with just one jvp and vjp instead of two, and maybe run both on a single core for a fair comparison (ideally disabling turbo boost and irqbalance; then running it with e.g. `taskset -c 2 ./test.py`).

Interestingly for our RBM toy example doing the jvp/vjp is actually quite a bit faster than precomputing the gradients.
@attila-i-szabo Would you be able to share some details on the kind of neural network used (number of parameters, samples, etc) ?

attila-i-szabo commented:
> Thanks for bringing this up :)
> The memory needed to store the forward when doing one big vjp (the way it is currently implemented in the master branch) is indeed something we were suspecting to be an issue (not the case for our RBM toy model though).

I wasn't saying this is an issue in itself, just that storing the gradient requires strictly less memory than performing a `vjp` which the lazy solution (as written now) needs to do.

> My proposed solution is to use the same approach already used for parallelizing it with MPI. The jacobian is split in blocks of rows which are multiplied from the left with a corresponding slice of the row vector. Then you loop over them while summing the results. (Taking it to the extreme you do just a batch of one sample at a time, which is just a vjp of a scalar function with a scalar, which is exactly what jax.grad is doing internally...). This way the memory needed can be reduced to that of a grad of a single sample at the expense of performance.
> 
> To anyone interested, feel free to take a look at [inailuig@0cd0b56](https://github.com/inailuig/netket/commit/0cd0b56408bafd43eee976e780e704c2280365fb) where I had a go at implementing this a couple of weeks ago (I also batched the jvp to see if it might improve compile times). It should already work for real parameters, but since jax scan/loops currently cannot be transposed (with AD; a requirement by custom_linear_solve in the iterative solvers) this is pending on me getting around to opening an issue/pr at jax...
> 
> Also there would be [jax.checkpoint](https://jax.readthedocs.io/en/latest/jax.html?highlight=jax.checkpoint#jax.checkpoint).
> 
> The main motivation for doing SR with jvp/vjp in the first place was that you also already cannot store the gradients (oks) when you have, say 100k parameters and samples.
> However your deep networks seem to be in the regime where precomputing the gradients is preferable, so this contribution is highly welcome!

This sounds like a useful extension -- at 100k parameters and samples, a naive `vjp` or `jvp` would also fail! However, I feel that before you hit this range, the precomputed option will usually perform better.

> > * It yields a massive speedup: the gradient matrix can again be calculated in the same asymptotic time as a single `vjp` (it will of course be a bit slower because `vmap` is used, but not by large factors); afterwards, we only need matrix-vector multiplications, which are way faster than a full backpropagation of a complex neural network. In my experiments (on 20
> >   CPUs), I could reduce the time of an SR step from 8x that with Adam to about 1.2x.
> 
> You could try `SR(..., centered=False)` which uses the implementation with just one jvp and vjp instead of two, and maybe run both on a single core for a fair comparison (ideally disabling turbo boost and irqbalance; then running it with e.g. `taskset -c 2 ./test.py`).

I'll try this when I have some spare computer time!

> Interestingly for our RBM toy example doing the jvp/vjp is actually quite a bit faster than precomputing the gradients.
> @attila-i-szabo Would you be able to share some details on the kind of neural network used (number of parameters, samples, etc) ?

That is very surprising given that precomputing the gradients has the same asymptotic runtime as one `vjp` and one would expect jit to remove most overhead prefactors. My guess is that backpropagating an RBM is so simple that you don't gain much by reducing it to a matrix multiplication, which is probably not the case for any deeper net.

My network has ~5000 parameters in 4 layers, wrapped with some untrained pre- and postprocessing layers. I trained it with 1024 samples. On the node I'm using, one Adam epoch took about 7 minutes, my SR code 7.5-8 minutes, the upstream SR code about an hour (I don't have very good statistics on this though, could be anywhere between 45-75 minutes).

PhilipVinc commented on netket/optimizer/sr/s_jacobian_mat.py:
```diff
@@ -0,0 +1,218 @@
+        sqrt_Skk = jnp.linalg.norm(grads, axis=0, keepdims=True)
+        return grads / sqrt_Skk, jnp.ravel(sqrt_Skk)
+    else:
+        return grads
```
```suggestion
        return grads, None
```

PhilipVinc commented on netket/optimizer/sr/sr_jacobian.py:
```diff
@@ -0,0 +1,142 @@
+    """Whether scale-invariant regularisation should be used"""
+
+    def create(self, *args, **kwargs):
+        if rescale_shift:
+            O, scale = gradients(apply_fun, params, samples, model_state, self.mode, True)
+            return JacobianSMatrix(sr=sr, x0=x0, O=O, scale=scale)
+        else:
+            O = gradients(apply_fun, params, samples, model_state, self.mode, False)
+            return JacobianSMatrix(sr=sr, x0=x0, O=O)
+
```
And with the change above this could became...
(I think it's a bit clearer)
```suggestion
        O, scale = gradients(apply_fun, params, samples, model_state, self.mode, rescale_shift)
        return JacobianSMatrix(sr=sr, x0=x0, O=O, scale=scale)
```

attila-i-szabo replied:
nice, this is much cleaner!

PhilipVinc commented on netket/optimizer/sr/s_jacobian_mat.py:
```diff
@@ -0,0 +1,218 @@
+            diag = jnp.eye(self.O.shape[1])
+        else:
+            O = self.O * self.scale[jnp.newaxis,:]
+            diag = jnp.diag(jnp.square(self.scale))
```
I prefer more clear code when possible and hate stuff like  `.square`.
In any case they compile down to the same thing
```python
>>> a=jax.numpy.ones(3)
>>> def f(x):
...   return jax.numpy.square(x)
...
>>> def g(x):
...   return x**2
...
>>> jax.make_jaxpr(f)(a)
{ lambda  ; a.
  let b = integer_pow[ y=2 ] a
  in (b,) }
>>> jax.make_jaxpr(g)(a)
{ lambda  ; a.
  let b = integer_pow[ y=2 ] a
  in (b,) }
```

```suggestion
            diag = jnp.diag(self.scale**2)
```

attila-i-szabo replied:
Cool. I was using `jnp.*` forms everywhere because I wasn't sure something like `x**2` or member functions would be evaluated on the GPU. But this is not a concern then?

PhilipVinc replied:
Jax first transforms code to IR (intermediate representation), what you see above, and then compiles it to a backend.
CPU/TPU is completely orthogonal concept

attila-i-szabo replied:
What I meant is that if the input is (say) a `np.ndarray`, then `x**2` would default to the numpy routine that would run on the CPU even if a GPU is available. But of course, we made sure these objects are `jnp.ndarray`s, so this isn't a concern...

PhilipVinc replied:
Ah I see.
It makes sense while prototyping but here we're so deep in the stack (and we're in a jitted block) so stuff is surely where you need it to be.

PhilipVinc commented on netket/optimizer/sr/s_jacobian_mat.py:
```diff
@@ -0,0 +1,218 @@
+            O = self.O * self.scale[jnp.newaxis,:]
+            diag = jnp.diag(jnp.square(self.scale))
+            
+        return jnp.transpose(jnp.conj(O)) @ O + self.sr.diag_shift * diag
```
```suggestion
        return O.T.conj() @ O + self.sr.diag_shift * diag
```

PhilipVinc replied:
mpi too

PhilipVinc commented on netket/optimizer/sr/s_jacobian_mat.py:
```diff
@@ -0,0 +1,218 @@
+        if self.scale is not None:
+            vec = vec * self.scale
+
+        result = jnp.transpose(jnp.conj(jnp.transpose(jnp.conj(self.O @ vec)) @ self.O)) + self.sr.diag_shift * vec
```
Can you split this equation in 2/3 lines for better readability?

Also, to support MPI I think you just have to put a `result = nk.stats.sum_inplace(result)` here before adding the shift

PhilipVinc commented on netket/optimizer/sr/s_jacobian_mat.py:
```diff
@@ -0,0 +1,218 @@
+
+    @jax.jit
+    def _unscaled_matmul(self, vec: jnp.ndarray) -> jnp.ndarray:
+        return jnp.transpose(jnp.conj(jnp.transpose(jnp.conj(S.O @ vec)) @ S.O)) + S.sr.diag_shift * vec
```
here too

PhilipVinc commented on netket/optimizer/sr/s_jacobian_mat.py:
```diff
@@ -0,0 +1,218 @@
+
+    if jnp.ndim(samples) != 2:
+        samples = jnp.reshape(samples, (-1, samples.shape[-1]))
+    n_samples = samples.shape[0]
```
```suggestion
    n_samples = samples.shape[0] * n_nodes
```

PhilipVinc commented on netket/optimizer/sr/s_jacobian_mat.py:
```diff
@@ -0,0 +1,218 @@
+    grads = jax.vmap(
+        jax.grad(fun, holomorphic=holomorphic), in_axes=(None, 0), out_axes=0
+    )(params, samples)
+    return grads - jnp.sum(grads, axis=0, keepdims=True) / grads.shape[0]
```
```suggestion
    return grads - sum_inplace(jnp.sum(grads, axis=0, keepdims=True)) / grads.shape[0]
```

PhilipVinc replied:
actually, you sum_inplace returns also a token object (second return value) that you should trash

attila-i-szabo replied:
I think it's taken care of [here](https://github.com/attila-i-szabo/netket/blob/8d9909e311be8d7860fc9661cce1eb3efd32b34c/netket/stats/_sum_inplace.py#L99)

PhilipVinc replied:
ah yes

PhilipVinc commented:
@attila-i-szabo what you are seeing is a very unfortunate consequence of bad decisions in python that will be fixed in python 3.10...
To fix it you can simply set a default value of None and in the `__post_init__` method error if that field is none.

attila-i-szabo commented:
Thanks, I'll have a go at that.

More importantly, I'm thinking of merging #652 and combining it into the `SRJacobian` class. To stop these ones from proliferating, I'd add a kwarg that specifies whether the `Oks` should be stored as a `"matrix"` or a `"pytree"` and return different `JacobianSMatrix` objects depending on it. For now, I'd stick to specifying the differentiation mode explicitly - there might be a way of overcoming the problem especially for the PyTrees, but I wouldn't treat it as a priority right now.

@inailuig @PhilipVinc should I go ahead with this?

PhilipVinc commented:
As i commented in the over PR, please don't.
I'd like to keep small, focused PR for better git history and to implement changes adiabatically when possible.
It leads to better reviewability too.
The point is also that I would like to get rid of all those keyword arguments in the SR structure as it is now, so that work you are proposing might be useless in the new design. Let's finalize the design and then do the UI/API work.

Does this sound good?

This PR is in a good state, once the MPI and datastructures issues are dealt with i want to merge it.
I'll also merge #652 when ready.

PhilipVinc commented:
Also, since you are a first time contributor, every time you send a commit GitHub nags me about checking your PR [in case you might be using Tests to mine cryptocurriencies](https://github.blog/2021-04-22-github-actions-update-helping-maintainers-combat-bad-actors/) and I have to approve it.

So if we get you merged you get your contributor badge and github stops harassing me. :-)

attila-i-szabo commented:
I'm working on fixing these bugs atm so it could be merged. When I install netket from this branch and try to import it in the interpreter, I get the following error:
```
>>> import netket as nk
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/attila/NQS/netket/netket/__init__.py", line 22, in <module>
    from ._version import version as __version__  # noqa: F401
ModuleNotFoundError: No module named 'netket._version'
```
Any ideas for how to fix that?

attila-i-szabo commented:
Ahh, thanks!

PhilipVinc commented:
it's because you did not run `pip install -e .`. That file is created when you run pip install

attila-i-szabo commented:
No, I was trying to run it from the directory of the local git repo and it was importing from `./netket` rather than the installed library. It's all good now. Also, I think the code should be good now, I could construct an `SRJacobian` object without errors at least. Is there anything missing before this can be merged?

PhilipVinc commented:
it looks like you should format your code with black.

PhilipVinc commented:
(pip install -U -e '.[dev]') to install the version we use 
you can also do in the netket folder `pre-commit install` to avoid committing code that won't pass those checks in the future

PhilipVinc commented:
If tests pass, it should be good.

attila-i-szabo commented:
Okay, the ubuntu test produces some weird error apparently unrelated to the SR code, so I'm not sure what to do with that. I've just put the code through black so I'll commit that.

PhilipVinc commented:
Yeah, that sampling test is a bit too sensible.
That looks good to me.
Il 28 apr 2021, 18:52 +0200, attila-i-szabo ***@***.***>, ha scritto:
> Okay, the ubuntu test produces some weird error apparently unrelated to the SR code, so I'm not sure what to do with that. I've just put the code through black so I'll commit that.
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

gcarleo commented:
but the formatting is still failing? make sure you are using the same black style we have maybe

attila-i-szabo commented:
I think you were still seeing the previous version, my most recent commit passed

PhilipVinc commented:
Yes that’s good. Indeed it passes test.
I’ll give a look later and eventually merge

PhilipVinc reviewed (changes requested):
No wait... did the test vanish?

PhilipVinc commented:
You just need to add a bunch of objects here https://github.com/netket/netket/blob/11bef1a9b03f482b3234f2fd0780e158a6d84637/Test/Optimizer/test_sr_itersolve.py#L26

PhilipVinc commented:
Pippo is now happy.
Pippo approves.

PhilipVinc commented:
Ah,
you should specify `mode=whatever`.
Ideally we would like to run another test for the three differentiation modes, but that means generalising the fixture creating the variational state...

attila-i-szabo commented:
Argh... the problem is that no single `mode` works for both complex and float RBMs, they need `holomorphic` and `R2R` respectively and would throw an error if used with other modes. Is there a way to supply a different set of SR objects depending on the dtype?

attila-i-szabo commented:
> If you don't want to do it let me know, I'll merge and take it from here.
> If you'd like to have fun writing tests I'll Giuseppe will owe you a beer if ever we are allowed back IRLconferences

I think I'd spare myself the fun for now :-) That code is too cryptic for me to actually interact with it!

On the other hand, @chrisrothUT found that `SRJacobianCG` errors out on one of his machines (which is based on flax, rather than stax, which I had tried). I'd wait with merging before that is resolved

Btw, this branch is now based on `netket:sr-types` rather than `master`. I wouldn't attempt to rebase it for it depends on the structure introduced in #650 too much for that. Would you merge that first and this on top?

PhilipVinc commented:
No my plan is to merge this into netket::sr-types, wait for the other PR, and then merge everything in master

PhilipVinc commented:
In Netket tests we test against flax machines. So we should catch errors

attila-i-szabo commented:
> No my plan is to merge this into netket::sr-types, wait for the other PR, and then merge everything in master

Sounds good, I was just wondering where #652 fits on this timeline?

> In Netket tests we test against flax machines. So we should catch errors

So it's probably not because of flax... will need more details from Chris anyway, it'd just be nice to make sure it works against most machines we might throw at it.

attila-i-szabo commented:
We should've cared to make the tests... I missed when the calling sequence of `create` was changed to `create(self, vstate, **kwargs)` so it errors out on missing arguments. Should I open a new PR to fix this?

PhilipVinc commented:
no don't worry, I just hit it and am solving it now

attila-i-szabo commented:
Cool, thanks!

PhilipVinc commented:
```python
    def solve(self, y: PyTree, x0: Optional[PyTree] = None) -> PyTree:
        """
        Solve the linear system x=⟨S⟩⁻¹⟨y⟩ with the chosen iterataive solver.

        Args:
            y: the vector y in the system above.
            x0: optional initial guess for the solution.

        Returns:
            x: the PyTree solving the system.
            info: optional additional informations provided by the solver. Might be
                None if there are no additional informations provided.
        """

        # Ravel input PyTrees, record unravelling function too
>       grad, unravel = nkjax.tree_ravel(grad)
E       UnboundLocalError: local variable 'grad' referenced before assignment
```

PhilipVinc commented:
that `grad` should be `y`, right?

attila-i-szabo commented:
Yup, copied too quickly...

PhilipVinc commented:
good. tests pass now.

attila-i-szabo commented:
Great, thanks for the fixes and sorry about the sloppy copying!
